File: sample-java/src/test/java/ru/solrudev/ackpine/sample/uninstall/UninstallViewModelTest.java
/*
 * Copyright (C) 2026 Ilya Fomichev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ru.solrudev.ackpine.sample.uninstall;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

import androidx.annotation.NonNull;
import androidx.arch.core.executor.testing.InstantTaskExecutorRule;
import androidx.lifecycle.SavedStateHandle;

import com.google.common.util.concurrent.MoreExecutors;

import org.junit.Rule;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ExecutorService;

import ru.solrudev.ackpine.session.Session;
import ru.solrudev.ackpine.test.TestSession;
import ru.solrudev.ackpine.test.TestSessionScript;
import ru.solrudev.ackpine.test.TestPackageUninstaller;
import ru.solrudev.ackpine.uninstaller.UninstallFailure;

public class UninstallViewModelTest {

	@Rule
	public final InstantTaskExecutorRule instantTaskExecutorRule = new InstantTaskExecutorRule();

	private final ExecutorService directExecutor = MoreExecutors.newDirectExecutorService();
	private static final String SESSION_ID_KEY = "SESSION_ID";
	private static final String PACKAGE_NAME_KEY = "PACKAGE_NAME";
	private static final String PACKAGE_NAME = "com.example.app";

	@Test
	public void loadApplicationsPopulatesUiState() {
		final var viewModel = new UninstallViewModel(new TestPackageUninstaller(),
				new SavedStateHandle(),
				directExecutor);
		final var app = createApplicationData();

		final var loadingValues = new ArrayList<Boolean>();
		viewModel.isLoading().observeForever(loadingValues::add);

		viewModel.loadApplications(true, () -> List.of(app));

		assertEquals(List.of(false, true, false), loadingValues);
		assertEquals(List.of(app), viewModel.getApplications().getValue());
	}

	@Test
	public void uninstallPackageSuccessfulFlow() {
		final var uninstaller = new TestPackageUninstaller();
		final var savedStateHandle = new SavedStateHandle();

		final var sessionIds = new ArrayList<UUID>();
		savedStateHandle.<UUID>getLiveData(SESSION_ID_KEY, null).observeForever(sessionIds::add);
		final var packageNames = new ArrayList<String>();
		savedStateHandle.<String>getLiveData(PACKAGE_NAME_KEY, null).observeForever(packageNames::add);

		final var viewModel = new UninstallViewModel(uninstaller, savedStateHandle, directExecutor);
		final var app = createApplicationData();
		viewModel.loadApplications(true, () -> List.of(app));

		viewModel.uninstallPackage(app.packageName());
		sessionIds.add(savedStateHandle.get(SESSION_ID_KEY));
		packageNames.add(savedStateHandle.get(PACKAGE_NAME_KEY));

		final var session = uninstaller.getSessions().getLast();
		session.getController().succeed();

		assertTrue(viewModel.getApplications().getValue().isEmpty());
		assertNull(viewModel.getFailure().getValue());
		assertEquals(Arrays.asList(null, session.getId(), null), sessionIds);
		assertEquals(Arrays.asList(null, app.packageName(), null), packageNames);
	}

	@Test
	public void uninstallPackageFailureFlow() {
		final var failure = new Session.State.Failed<>(new UninstallFailure.Generic("Failure"));
		TestSessionScript<UninstallFailure> script = TestSessionScript.auto(failure);
		final var uninstaller = new TestPackageUninstaller(script);
		final var savedStateHandle = new SavedStateHandle();
		final var viewModel = new UninstallViewModel(uninstaller, savedStateHandle, directExecutor);
		final var app = createApplicationData();

		viewModel.loadApplications(true, () -> List.of(app));
		viewModel.uninstallPackage(app.packageName());
		uninstaller.getSessions().getLast().getController().fail(new UninstallFailure.Generic("Failure"));

		assertEquals("Failure", viewModel.getFailure().getValue());
		assertEquals(List.of(app), viewModel.getApplications().getValue());
		assertNull(savedStateHandle.get(SESSION_ID_KEY));
		assertNull(savedStateHandle.get(PACKAGE_NAME_KEY));

		viewModel.clearFailure();
		assertNull(viewModel.getFailure().getValue());
	}

	@Test
	public void restoresSessionFromSavedStateAndClearsOnCompletion() {
		final var session = new TestSession<UninstallFailure>();
		final var uninstaller = new TestPackageUninstaller();
		uninstaller.seedSession(session);
		final var savedStateHandle = new SavedStateHandle(
				Map.of(
						SESSION_ID_KEY, session.getId(),
						PACKAGE_NAME_KEY, PACKAGE_NAME
				)
		);

		new UninstallViewModel(uninstaller, savedStateHandle, directExecutor);

		assertNull(savedStateHandle.get(SESSION_ID_KEY));
		assertNull(savedStateHandle.get(PACKAGE_NAME_KEY));
	}

	@NonNull
	private static ApplicationData createApplicationData() {
		return new ApplicationData("App", PACKAGE_NAME, new TestDrawable());
	}
}

File: sample-ktx/src/test/kotlin/ru/solrudev/ackpine/sample/uninstall/UninstallViewModelTest.kt
/*
 * Copyright (C) 2026 Ilya Fomichev
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ru.solrudev.ackpine.sample.uninstall

import androidx.lifecycle.SavedStateHandle
import app.cash.turbine.test
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.Rule
import ru.solrudev.ackpine.libsu.LibsuPlugin
import ru.solrudev.ackpine.sample.MainDispatcherRule
import ru.solrudev.ackpine.sample.settings.InstallerBackend
import ru.solrudev.ackpine.sample.settings.createSettingsRepository
import ru.solrudev.ackpine.session.Session
import ru.solrudev.ackpine.shizuku.ShizukuPlugin
import ru.solrudev.ackpine.test.TestPackageUninstaller
import ru.solrudev.ackpine.test.TestSessionScript
import ru.solrudev.ackpine.test.TestUninstallSession
import ru.solrudev.ackpine.uninstaller.UninstallFailure
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

private const val SESSION_ID_KEY = "SESSION_ID"
private const val PACKAGE_NAME_KEY = "PACKAGE_NAME"
private const val PACKAGE_NAME = "com.example.app"

@OptIn(ExperimentalCoroutinesApi::class)
class UninstallViewModelTest {

	@JvmField
	@Rule
	val mainDispatcherRule = MainDispatcherRule()

	@Test
	fun loadApplicationsPopulatesUiState() = runTest(mainDispatcherRule.dispatcher) {
		val viewModel = UninstallViewModel(
			TestPackageUninstaller(),
			SavedStateHandle(),
			coroutineContext,
			createSettingsRepository()
		)
		val app = createApplicationData()
		viewModel.uiState.test {
			awaitItem() // initial state

			viewModel.loadApplications(refresh = true) { listOf(app) }

			val expectedState1 = UninstallUiState(isLoading = true)
			assertEquals(expectedState1, awaitItem())

			val expectedState2 = UninstallUiState(
				isLoading = false,
				applications = listOf(app)
			)
			assertEquals(expectedState2, awaitItem())
		}
	}

	@Test
	fun uninstallPackageSuccessfulFlow() = runTest(mainDispatcherRule.dispatcher) {
		val uninstaller = TestPackageUninstaller(TestSessionScript.empty())
		val savedStateHandle = SavedStateHandle()
		val viewModel = UninstallViewModel(uninstaller, savedStateHandle, coroutineContext, createSettingsRepository())
		val app = createApplicationData()
		viewModel.uiState.test {
			awaitItem() // initial state
			viewModel.loadApplications(refresh = true) { listOf(app) }
			awaitItem() // loading apps
			awaitItem() // apps loaded

			viewModel.uninstallPackage(app.packageName)
			advanceUntilIdle()
			val session = uninstaller.sessions.last()
			assertEquals(session.id, savedStateHandle[SESSION_ID_KEY])
			assertEquals(app.packageName, savedStateHandle[PACKAGE_NAME_KEY])

			session.controller.succeed()
			assertEquals(UninstallUiState(), awaitItem())
			assertNull(savedStateHandle[SESSION_ID_KEY])
			assertNull(savedStateHandle[PACKAGE_NAME_KEY])
		}
	}

	@Test
	fun uninstallPackageFailureFlow() = testUninstallFailure(
		failure = UninstallFailure.Generic("Failure"),
		expectedError = "Failure"
	)

	@Test
	fun uninstallPackageExceptionFlow() = testUninstallFailure(
		failure = UninstallFailure.Exceptional(Exception("exception")),
		expectedError = "exception"
	)

	private fun testUninstallFailure(
		failure: UninstallFailure,
		expectedError: String
	) = runTest(mainDispatcherRule.dispatcher) {
		val script: TestSessionScript<UninstallFailure> = TestSessionScript.auto(Session.State.Failed(failure))
		val uninstaller = TestPackageUninstaller(script)
		val savedStateHandle = SavedStateHandle()
		val viewModel = UninstallViewModel(uninstaller, savedStateHandle, coroutineContext, createSettingsRepository())
		val app = createApplicationData()
		viewModel.uiState.test {
			awaitItem() // initial state
			viewModel.loadApplications(refresh = true) { listOf(app) }
			awaitItem() // loading apps
			awaitItem() // apps loaded

			viewModel.uninstallPackage(app.packageName)

			val expectedState = UninstallUiState(
				isLoading = false,
				applications = listOf(app),
				failure = expectedError
			)
			assertEquals(expectedState, awaitItem())
			assertNull(savedStateHandle[SESSION_ID_KEY])
			assertNull(savedStateHandle[PACKAGE_NAME_KEY])

			viewModel.clearFailure()
			assertEquals(expectedState.copy(failure = null), awaitItem())
		}
	}

	@Test
	fun restoresSessionFromSavedStateAndClearsOnCompletion() = runTest(mainDispatcherRule.dispatcher) {
		val script: TestSessionScript<UninstallFailure> = TestSessionScript.auto(Session.State.Succeeded)
		val sessionId = UUID.randomUUID()
		val session = TestUninstallSession(script, sessionId)
		val uninstaller = TestPackageUninstaller()
		uninstaller.seedSession(session)
		val savedStateHandle = SavedStateHandle(
			mapOf(
				SESSION_ID_KEY to sessionId,
				PACKAGE_NAME_KEY to PACKAGE_NAME
			)
		)
		val viewModel = UninstallViewModel(uninstaller, savedStateHandle, coroutineContext, createSettingsRepository())
		viewModel.uiState.first() // trigger collection
		advanceUntilIdle()

		assertNull(savedStateHandle[SESSION_ID_KEY])
		assertNull(savedStateHandle[PACKAGE_NAME_KEY])
	}

	@Test
	fun uninstallPackageRegistersLibsuPluginForRootBackend() = runTest(mainDispatcherRule.dispatcher) {
		val uninstaller = TestPackageUninstaller(TestSessionScript.empty())
		val savedStateHandle = SavedStateHandle()
		val settingsRepository = createSettingsRepository(InstallerBackend.ROOT)
		val viewModel = UninstallViewModel(uninstaller, savedStateHandle, coroutineContext, settingsRepository)

		viewModel.uninstallPackage(PACKAGE_NAME)
		advanceUntilIdle()

		val session = uninstaller.sessions.last()
		val parameters = uninstaller.createdParameters.getValue(session.id)
		val pluginParameters = parameters.pluginContainer
			.getPlugins()
			.getValue(LibsuPlugin::class.java) as LibsuPlugin.UninstallParameters
		assertEquals(LibsuPlugin.UninstallParameters.DEFAULT, pluginParameters)
	}

	@Test
	fun uninstallPackageRegistersShizukuPluginForShizukuBackend() = runTest(mainDispatcherRule.dispatcher) {
		val uninstaller = TestPackageUninstaller(TestSessionScript.empty())
		val savedStateHandle = SavedStateHandle()
		val settingsRepository = createSettingsRepository(InstallerBackend.SHIZUKU, supportsShizuku = flowOf(true))
		val viewModel = UninstallViewModel(uninstaller, savedStateHandle, coroutineContext, settingsRepository)

		viewModel.uninstallPackage(PACKAGE_NAME)
		advanceUntilIdle()

		val session = uninstaller.sessions.last()
		val parameters = uninstaller.createdParameters.getValue(session.id)
		val pluginParameters = parameters.pluginContainer
			.getPlugins()
			.getValue(ShizukuPlugin::class.java) as ShizukuPlugin.UninstallParameters
		assertEquals(ShizukuPlugin.UninstallParameters.DEFAULT, pluginParameters)
	}

	private fun createApplicationData() = ApplicationData("App", PACKAGE_NAME, TestDrawable())
}